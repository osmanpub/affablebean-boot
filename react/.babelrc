 
{
  "presets": ["@babel/preset-react"],
  "env": {
    "test": {
      "plugins": [
        ["istanbul", {
          "exclude": [
            "**/*.spec.js"
          ]
        }]
      ]
    }
  },
  "plugins": ["transform-class-properties", "istanbul"]
}
