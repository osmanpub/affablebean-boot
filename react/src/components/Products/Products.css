.categoryButton {
  margin: 15px 22px;
  padding: 13px;
  display: block;
  background-color: #d3ede8;
  border-radius: 4px;
}

.categoryText {
  line-height: 25px;
  font-size: x-large;
}
