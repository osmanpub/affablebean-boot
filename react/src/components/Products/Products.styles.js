import styled from "styled-components";

export const CategoriesGreeting = styled.p`
  font-size larger;
`;

// #productTable {
// 	width: 600px;

// 	/* styles to collaborate with rounded categoryLabel */
// 	position: relative;
// 	z-index: 1;
// }

// #productTable tr { height: 90px }

// #productTable td { width: 145px }

// #categoryLeftColumn {
// 	width: 185px;
// 	float: left;
// 	margin-top: 25px;
// 	padding-left: 15px;
// }

// #categoryRightColumn {
// 	margin-top: 10px;
// 	width: 650px;
// 	float: left;
// }

// #selectedCategory {
// 	background-color: #b2d2d2;
// 	margin-left: 10px;
// 	width: 139px;
// }

// #categoryTitle {
// 	margin-left: 700px;
// 	margin-top: -40px;
// 	/*right: 200px;*/
// 	font-size: x-large;
// 	/*position: absolute;*/
// 	background-color: #f5eabe;
// 	/*padding: 7px;*/
// 	/*border-radius: 4px;*/
// }
