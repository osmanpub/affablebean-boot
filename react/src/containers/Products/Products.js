import React, { Component } from "react";
import { connect } from "react-redux";
import Categories from "../../components/Categories";
import { fetchCategoryIfNeeded } from "../../rest/category";
import Footer from "../../components/Footer";
import Header from "../../components/Header";

export class Products extends Component {
  componentDidMount() {
    const { dispatch } = this.props;    
    dispatch(fetchCategoryIfNeeded());
  }

  // 3 renders on start is too much - investigate
  render() {
    const { categories, match }= this.props;

    return (
        <Header />
        {!match && <Categories categories={categories.items} />}
        <Footer />
    );
  }
}

const mapStateToProps = state => {
  const { categories } = state;

  return {
    categories
  };
};

export default connect(mapStateToProps)(App);
