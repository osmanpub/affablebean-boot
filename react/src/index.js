import React from "react";
import { render } from "react-dom";
import { configureStore } from "redux-starter-kit";
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route } from "react-router-dom";
import reducer from "./reducers";
import App from "./containers/App";
import CategoryProducts from "./containers/CategoryProducts";
import * as serviceWorker from "./serviceWorker";

const store = configureStore({
  reducer
});

render(
  // <React.StrictMode>
  <Provider store={store}>
    <Router>
      <Route path="/" exact component={App} />
      <Route path="/category/:id" component={CategoryProducts} />
    </Router>
  </Provider>,
  // </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
serviceWorker.unregister();
