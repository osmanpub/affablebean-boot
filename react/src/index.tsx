import { configureStore } from "@reduxjs/toolkit";
import React, { lazy, Suspense } from "react";
import { render } from "react-dom";
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import ErrorBoundary from "./components/ErrorBoundary";
import rootReducer from "./redux";
import * as serviceWorker from "./serviceWorker";
const Cart = lazy(() => import("./containers/Cart"));
const CategoryProducts = lazy(() => import("./containers/CategoryProducts"));
const Checkout = lazy(() => import("./containers/Checkout"));
const Contact = lazy(() => import("./containers/Contact"));
const Home = lazy(() => import("./containers/Home"));
const Privacy = lazy(() => import("./containers/Privacy"));

const store = configureStore({
  reducer: rootReducer
});

render(
  <React.StrictMode>
    <ErrorBoundary>
      <Provider store={store}>
        <Router>
          <Suspense fallback={<div>Loading...</div>}>
            <Switch>
              <Route path="/" exact component={Home} />
              <Route path="/category/:id" component={CategoryProducts} />
              <Route path="/checkout" component={Checkout} />
              <Route path="/contact" component={Contact} />
              <Route path="/privacy" component={Privacy} />
              <Route path="/viewCart/:clear" component={Cart} />
            </Switch>
          </Suspense>
        </Router>
      </Provider>
    </ErrorBoundary>
  </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
serviceWorker.unregister();
