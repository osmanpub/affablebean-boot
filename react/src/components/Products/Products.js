import React, { Component } from "react";
import { Link } from "react-router-dom";
import "./Products.css";
import { addProductToCart } from "../../rest/cart";
import {
  CategoryTitle,
  ProductsLeft,
  ProductsRight,
  ProductsTable,
  SelectedCategory
} from "./Products.styles";

export class Products extends Component {
  constructor(props) {
    super(props);
    this.addToCart = this.addToCart.bind(this);
  }

  addToCart(id) {
    const { dispatch } = this.props;
    dispatch(addProductToCart(id));
  }

  render() {
    const categories = this.props.categories;

    if (categories.length === 0) {
      return null;
    }

    const selectedCategory = this.props.category;

    const sidePanel = categories.map(category => {
      const key = category._links.self.href;
      const name = category.name;

      if (name === selectedCategory.name) {
        return (
          <SelectedCategory key={key}>
            <span className="categoryText">{name}</span>
          </SelectedCategory>
        );
      } else {
        return (
          <Link key={key} to={"/category/" + category.id}>
            <span className="categoryButton">
              <span className="categoryText">{name}</span>
            </span>
          </Link>
        );
      }
    });

    const products = this.props.products._embedded.productList.map(
      (product, index) => {
        const name = product.name;

        return (
          <tr
            key={product._links.self.href}
            className="{index % 2 === 0 ? 'white' : 'lightBlue'}"
          >
            <td>
              <img
                src={"/static/img/products/" + name + ".png"}
                alt="{{product.name}"
              />
            </td>

            <td>
              <span>{name}</span>
              <br />
              <span className="smallText">{product.description}</span>
            </td>

            <td>
              &euro;&nbsp;
              {product.price}
            </td>

            <td>&nbsp;&nbsp;</td>

            <td>
              <button
                className="btn btn-primary btn-sm"
                onClick={this.addToCart.bind(this, product.id)}
              >
                <span>add</span>
              </button>
            </td>

            <td>&nbsp;&nbsp;</td>
          </tr>
        );
      }
    );

    return (
      <div>
        <ProductsLeft>{sidePanel}</ProductsLeft>
        <ProductsRight>
          <CategoryTitle>{selectedCategory.name}</CategoryTitle>
          <ProductsTable>
            <tbody>{products}</tbody>
          </ProductsTable>
        </ProductsRight>
      </div>
    );
  }
}
