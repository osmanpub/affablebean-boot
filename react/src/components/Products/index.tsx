import React from "react";
import { connect, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { getId } from "../../helpers/utils";
import { Cart } from "../../interfaces/cart";
import {
  Category,
  CategoryState,
  ProductState
} from "../../interfaces/categories";
import { id } from "../../interfaces/id";
import { addProductToCart, updateProductInCart } from "../../net/cart";
import { RootState } from "../../redux";
import { clearPurchase } from "../../redux/purchase";
import "./Products.css";
import {
  CategoryTitle,
  ProductsLeft,
  ProductsRight,
  ProductsTable,
  SelectedCategory
} from "./Products.styles";

type Props = {
  cart: Cart;
  categories: Array<CategoryState>;
  category: Category;
  clearPurchase: Function;
  products: Array<ProductState>;
};

function Products(props: Props) {
  const { cart, categories, clearPurchase, products } = props;
  const dispatch = useDispatch();

  const addToCart = (id: id) => {
    const update = cart.items.filter(item => getId(item.product) === id);

    clearPurchase();
    dispatch(
      update.length > 0
        ? updateProductInCart(id, update[0].quantity + 1)
        : addProductToCart(id.toString())
    );
  };

  if (categories.length === 0) {
    return null;
  }

  const selectedCategory = props.category;

  const sidePanel = categories.map((category: CategoryState) => {
    const id = getId(category);
    const name = category.name;

    if (name === selectedCategory.name) {
      return (
        <SelectedCategory key={id}>
          <span className="categoryText">{name}</span>
        </SelectedCategory>
      );
    } else {
      return (
        <Link key={id} to={`/category/${id}`}>
          <span className="categoryButton">
            <span className="categoryText" data-cy={`category-${name}`}>
              {name}
            </span>
          </span>
        </Link>
      );
    }
  });

  const productsList = products.map((product: ProductState, index: number) => {
    const id = getId(product);
    const name = product.name;
    const rowCol = index % 2 === 0 ? "white" : "lightBlue";

    return (
      <tr key={id} className={`${rowCol}`} data-cy={`product-${name}`}>
        <td>
          <img src={`/static/img/products/${name}.png`} alt="{name}" />
        </td>

        <td>
          {name}
          <br />
          <span className="smallText">{product.description}</span>
        </td>

        <td>
          &euro;&nbsp;
          {product.price.toFixed(2)}
        </td>

        <td>&nbsp;&nbsp;</td>

        <td>
          <button
            className="btn btn-primary btn-sm"
            onClick={() => addToCart(id)}
          >
            add
          </button>
        </td>

        <td>&nbsp;&nbsp;</td>
      </tr>
    );
  });

  return (
    <div>
      <ProductsLeft>{sidePanel}</ProductsLeft>
      <ProductsRight>
        <CategoryTitle>{selectedCategory.name}</CategoryTitle>
        <ProductsTable>
          <tbody>{productsList}</tbody>
        </ProductsTable>
      </ProductsRight>
    </div>
  );
}

const mapStateToProps = (state: RootState) => ({
  cart: state.cart
});

const mapDispatchToProps = {
  clearPurchase
};

export default connect(mapStateToProps, mapDispatchToProps)(Products);
