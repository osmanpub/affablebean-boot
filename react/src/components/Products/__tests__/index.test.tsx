import { render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import Products from "..";

const mockStore = configureStore([]);

const initialState = {
  cart: {
    didInvalidate: false,
    isFetching: false,
    items: [],
    numberOfItems: 0,
    subtotal: 0
  }
};

const store = mockStore(initialState);

const category = {
  categories: [
    {
      id: 3,
      name: "bakery",
      _links: {
        self: { href: "http://localhost:8080/api/categories/3" },
        categories: { href: "http://localhost:8080/api/categories" }
      }
    },
    {
      id: 5,
      name: "cereals",
      _links: {
        self: { href: "http://localhost:8080/api/categories/5" },
        categories: { href: "http://localhost:8080/api/categories" }
      }
    },
    {
      id: 1,
      name: "dairy",
      _links: {
        self: { href: "http://localhost:8080/api/categories/1" },
        categories: { href: "http://localhost:8080/api/categories" }
      }
    },
    {
      id: 6,
      name: "drinks",
      _links: {
        self: { href: "http://localhost:8080/api/categories/6" },
        categories: { href: "http://localhost:8080/api/categories" }
      }
    },
    {
      id: 4,
      name: "fruit & veg",
      _links: {
        self: { href: "http://localhost:8080/api/categories/4" },
        categories: { href: "http://localhost:8080/api/categories" }
      }
    },
    {
      id: 2,
      name: "meats",
      _links: {
        self: { href: "http://localhost:8080/api/categories/2" },
        categories: { href: "http://localhost:8080/api/categories" }
      }
    }
  ],
  category: {
    id: 1,
    name: "dairy",
    _links: {
      self: { href: "http://localhost:8080/api/categories/1" },
      categories: { href: "http://localhost:8080/api/categories" }
    }
  },
  products: [
    // node
    {
      id: 1,
      description: "semi skimmed (1L)",
      name: "milk",
      price: 1.7,
      category: { id: 1, name: "dairy" },
      _links: {
        self: { href: "http://localhost:8080/api/products/1" },
        products: { href: "http://localhost:8080/api/products" }
      }
    },
    {
      id: 2,
      description: "mild cheddar (330g)",
      name: "cheese",
      price: 2.39,
      category: { id: 1, name: "dairy" },
      _links: {
        self: { href: "http://localhost:8080/api/products/2" },
        products: { href: "http://localhost:8080/api/products" }
      }
    },
    {
      id: 3,
      description: "unsalted (250g)",
      name: "butter",
      price: 1.09,
      category: { id: 1, name: "dairy" },
      _links: {
        self: { href: "http://localhost:8080/api/products/3" },
        products: { href: "http://localhost:8080/api/products" }
      }
    },
    {
      id: 4,
      description: "medium-sized (6 eggs)",
      name: "free range eggs",
      price: 1.76,
      category: { id: 1, name: "dairy" },
      _links: {
        self: { href: "http://localhost:8080/api/products/4" },
        products: { href: "http://localhost:8080/api/products" }
      }
    }
  ]
  // products: { // java
  //   _embedded: {
  //     productList: [
  //       {
  //         id: 1,
  //         description: "semi skimmed (1L)",
  //         name: "milk",
  //         price: 1.7,
  //         category: { id: 1, name: "dairy" },
  //         _links: {
  //           self: { href: "http://localhost:8080/api/products/1" },
  //           products: { href: "http://localhost:8080/api/products" }
  //         }
  //       },
  //       {
  //         id: 2,
  //         description: "mild cheddar (330g)",
  //         name: "cheese",
  //         price: 2.39,
  //         category: { id: 1, name: "dairy" },
  //         _links: {
  //           self: { href: "http://localhost:8080/api/products/2" },
  //           products: { href: "http://localhost:8080/api/products" }
  //         }
  //       },
  //       {
  //         id: 3,
  //         description: "unsalted (250g)",
  //         name: "butter",
  //         price: 1.09,
  //         category: { id: 1, name: "dairy" },
  //         _links: {
  //           self: { href: "http://localhost:8080/api/products/3" },
  //           products: { href: "http://localhost:8080/api/products" }
  //         }
  //       },
  //       {
  //         id: 4,
  //         description: "medium-sized (6 eggs)",
  //         name: "free range eggs",
  //         price: 1.76,
  //         category: { id: 1, name: "dairy" },
  //         _links: {
  //           self: { href: "http://localhost:8080/api/products/4" },
  //           products: { href: "http://localhost:8080/api/products" }
  //         }
  //       }
  //     ]
  //   },
  //   _links: {
  //     categoryProducts: { href: "http://localhost:8080/api/products" }
  //   }
  // }
};

const products = (
  <Router>
    <Provider store={store}>
      <Products {...category} />
    </Provider>
  </Router>
);

describe("<Products />", () => {
  it("renders correctly", () => {
    const component = renderer.create(products).toJSON();
    // @ts-ignore
    expect(component).toMatchSnapshot();
  });

  it("selected category is dairy", () => {
    const { getByTestId } = render(products);
    const selected = getByTestId(/selected-dairy/i);
    // @ts-ignore
    expect(selected).toHaveTextContent("dairy");
  });

  it("four products are shown", () => {
    const { getAllByTestId } = render(products);
    const items = getAllByTestId(/product/i);
    // @ts-ignore
    expect(items).toHaveLength(4);
  });

  it("show submit buttons", () => {
    const { getAllByRole } = render(products);
    const buttons = getAllByRole(/button/i);
    // @ts-ignore
    expect(buttons).toHaveLength(4);
  });

  it("first product is 'milk'", () => {
    const { getByAltText, getByTestId, getAllByTestId } = render(products);
    const items = getAllByTestId(/product/i);
    const milk = items[0];
    // @ts-ignore
    expect(milk).toHaveTextContent("milk");
    // @ts-ignore
    expect(milk).toHaveTextContent("1.7");
    // @ts-ignore
    expect(milk).toHaveTextContent("semi skimmed (1L)");

    const img = getByAltText(/milk/i);
    // @ts-ignore
    expect(img).toBeInTheDocument();

    const price = getByTestId(/price-milk/i);
    // @ts-ignore
    expect(price).toHaveTextContent("1.7");
  });
});
