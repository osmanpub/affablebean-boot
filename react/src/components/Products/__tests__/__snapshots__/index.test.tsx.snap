// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products /> renders correctly 1`] = `
<div>
  <div
    className="sc-fznZeY iUOeJ"
  >
    <a
      href="/category/3"
      onClick={[Function]}
    >
      <span
        className="categoryButton"
      >
        <span
          className="categoryText"
          data-cy="category-bakery"
          data-testid="category-bakery"
        >
          bakery
        </span>
      </span>
    </a>
    <a
      href="/category/5"
      onClick={[Function]}
    >
      <span
        className="categoryButton"
      >
        <span
          className="categoryText"
          data-cy="category-cereals"
          data-testid="category-cereals"
        >
          cereals
        </span>
      </span>
    </a>
    <span
      className="sc-fzqNJr ddDkXI"
      data-testid="selected-dairy"
    >
      <span
        className="categoryText"
      >
        dairy
      </span>
    </span>
    <a
      href="/category/6"
      onClick={[Function]}
    >
      <span
        className="categoryButton"
      >
        <span
          className="categoryText"
          data-cy="category-drinks"
          data-testid="category-drinks"
        >
          drinks
        </span>
      </span>
    </a>
    <a
      href="/category/4"
      onClick={[Function]}
    >
      <span
        className="categoryButton"
      >
        <span
          className="categoryText"
          data-cy="category-fruit & veg"
          data-testid="category-fruit & veg"
        >
          fruit & veg
        </span>
      </span>
    </a>
    <a
      href="/category/2"
      onClick={[Function]}
    >
      <span
        className="categoryButton"
      >
        <span
          className="categoryText"
          data-cy="category-meats"
          data-testid="category-meats"
        >
          meats
        </span>
      </span>
    </a>
  </div>
  <div
    className="sc-fzokOt jfgkMv"
  >
    <p
      className="sc-fznKkj fQnMFj"
    >
      dairy
    </p>
    <table
      className="sc-fzqBZW igCcce"
    >
      <tbody>
        <tr
          className="white"
          data-cy="product-milk"
          data-testid="product-milk"
        >
          <td>
            <img
              alt="milk"
              src="/static/img/products/milk.png"
            />
          </td>
          <td
            data-testid="name"
          >
            milk
            <br />
            <span
              className="smallText"
            >
              semi skimmed (1L)
            </span>
          </td>
          <td
            data-testid="price"
          >
            € 
            1.70
          </td>
          <td>
              
          </td>
          <td>
            <button
              className="btn btn-primary btn-sm"
              onClick={[Function]}
            >
              add
            </button>
          </td>
          <td>
              
          </td>
        </tr>
        <tr
          className="lightBlue"
          data-cy="product-cheese"
          data-testid="product-cheese"
        >
          <td>
            <img
              alt="cheese"
              src="/static/img/products/cheese.png"
            />
          </td>
          <td
            data-testid="name"
          >
            cheese
            <br />
            <span
              className="smallText"
            >
              mild cheddar (330g)
            </span>
          </td>
          <td
            data-testid="price"
          >
            € 
            2.39
          </td>
          <td>
              
          </td>
          <td>
            <button
              className="btn btn-primary btn-sm"
              onClick={[Function]}
            >
              add
            </button>
          </td>
          <td>
              
          </td>
        </tr>
        <tr
          className="white"
          data-cy="product-butter"
          data-testid="product-butter"
        >
          <td>
            <img
              alt="butter"
              src="/static/img/products/butter.png"
            />
          </td>
          <td
            data-testid="name"
          >
            butter
            <br />
            <span
              className="smallText"
            >
              unsalted (250g)
            </span>
          </td>
          <td
            data-testid="price"
          >
            € 
            1.09
          </td>
          <td>
              
          </td>
          <td>
            <button
              className="btn btn-primary btn-sm"
              onClick={[Function]}
            >
              add
            </button>
          </td>
          <td>
              
          </td>
        </tr>
        <tr
          className="lightBlue"
          data-cy="product-free range eggs"
          data-testid="product-free range eggs"
        >
          <td>
            <img
              alt="free range eggs"
              src="/static/img/products/free range eggs.png"
            />
          </td>
          <td
            data-testid="name"
          >
            free range eggs
            <br />
            <span
              className="smallText"
            >
              medium-sized (6 eggs)
            </span>
          </td>
          <td
            data-testid="price"
          >
            € 
            1.76
          </td>
          <td>
              
          </td>
          <td>
            <button
              className="btn btn-primary btn-sm"
              onClick={[Function]}
            >
              add
            </button>
          </td>
          <td>
              
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
